File: telegram/src/main/java/by/mrrockka/service/game/GameTelegramService.java
package by.mrrockka.service.game;

import by.mrrockka.domain.MessageMetadata;
import by.mrrockka.domain.TelegramGame;
import by.mrrockka.mapper.game.TelegramGameMapper;
import by.mrrockka.repo.game.TelegramGameRepository;
import by.mrrockka.service.GameService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.telegram.telegrambots.meta.api.methods.botapimethods.BotApiMethodMessage;
import org.telegram.telegrambots.meta.api.objects.Update;

import java.util.Optional;

@Slf4j
@Service
@RequiredArgsConstructor
public class GameTelegramService {

  private final TelegramGameRepository telegramGameRepository;
  private final TournamentGameService tournamentGameService;
  private final CashGameService cashGameService;
  private final BountyGameService bountyGameService;
  private final GameService gameService;
  private final TelegramGameMapper telegramGameMapper;

  public BotApiMethodMessage storeTournamentGame(final Update update) {
    return tournamentGameService.storeGame(update);
  }

  public BotApiMethodMessage storeCashGame(final Update update) {
    return cashGameService.storeGame(update);
  }

  public BotApiMethodMessage storeBountyGame(final Update update) {
    return bountyGameService.storeGame(update);
  }

  public Optional<TelegramGame> getGameByMessageMetadata(final MessageMetadata messageMetadata) {
    return messageMetadata.optReplyTo()
      .map(replyTo -> telegramGameRepository.findByChatAndMessageId(messageMetadata.chatId(), replyTo.id()))
      .orElseGet(() -> telegramGameRepository.findLatestByChatId(messageMetadata.chatId()))
      .map(entity -> telegramGameMapper.toGame(gameService.retrieveGame(entity.gameId()), entity));
  }

}

File: telegram/src/main/java/by/mrrockka/service/game/GameTelegramService.kt
package by.mrrockka.service.game

import by.mrrockka.domain.*
import by.mrrockka.parser.GameMessageParser
import by.mrrockka.repo.ChatGameRepo
import by.mrrockka.repo.EntriesRepo
import by.mrrockka.repo.GameRepo
import by.mrrockka.service.TelegramPersonService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.telegram.telegrambots.meta.api.methods.botapimethods.BotApiMethodMessage
import org.telegram.telegrambots.meta.api.methods.send.SendMessage

interface GameTelegramService {
    fun storeGame(messageMetadata: MessageMetadata): BotApiMethodMessage
    fun findGame(messageMetadata: MessageMetadata): ChatGame
}

@Service
@Transactional
open class GameTelegramServiceImpl(
        private val telegramPersonService: TelegramPersonService,
        private val gameRepo: GameRepo,
        private val entriesRepo: EntriesRepo,
        private val chatGameRepo: ChatGameRepo,
        private val gameMessageParser: GameMessageParser,
) : GameTelegramService {


    override fun storeGame(messageMetadata: MessageMetadata): BotApiMethodMessage {
        check(messageMetadata.mentions().isNotEmpty()) { "Game must have at least one player" }

        val game = gameMessageParser.parse(messageMetadata)
        gameRepo.save(game)
        val personIds = telegramPersonService.findByMentions(messageMetadata)
        entriesRepo.insertBatch(personIds, game, messageMetadata.createdAt)
        chatGameRepo.store(game.id, messageMetadata)

        return SendMessage().apply {
            chatId = messageMetadata.chatId.toString()
            text = game.responseMessage()
            replyToMessageId = messageMetadata.id
        }
    }

    override fun findGame(messageMetadata: MessageMetadata): ChatGame {
        val chatGameId = if (messageMetadata.replyTo != null) chatGameRepo.findByMessage(messageMetadata)
        else chatGameRepo.findLatestForChat(messageMetadata)

        check(chatGameId != null) { "Game was not found for the chat." }

        return ChatGame(game = gameRepo.findById(chatGameId), messageMetadata = messageMetadata)
    }

    private fun Game.responseMessage(): String =
            when (this) {
                is CashGame -> "Cash game started."
                is TournamentGame -> "Tournament game started."
                is BountyTournamentGame -> "Bounty tournament game started."
                else -> error("Game type not supported: ${this.javaClass.simpleName}")
            }

}
